src/array/cpu/spmm.cc

```diff
@@ -40,6 +40,60 @@ void SpMMCsr(const std::string& op, const std::string& reduce,
   }
 }
 
+/*! \brief Generalized SpMM on Csr format. */
+template <int XPU, typename IdType, int bits>
+void SpMMCsrHetero(const std::string& op, const std::string& reduce,
+             const BcastOff& bcast,
+             const std::vector<CSRMatrix>& vec_csr,
+             std::vector<NDArray> vec_ufeat,
+             std::vector<NDArray> vec_efeat,
+             std::vector<NDArray> vec_out,
+             std::vector<NDArray> out_aux,
+             const std::vector<dgl_type_t> ufeat_eid,
+             const std::vector<dgl_type_t> out_eid) {
+  if (reduce == "sum") {
+    SWITCH_BITS(bits, DType, {
+      SWITCH_OP(op, Op, {
+        /* Call  SpMM for each relation type */
+        for (dgl_type_t etype = 0; etype < ufeat_eid.size(); ++etype) {
+          const dgl_type_t src_id = ufeat_eid[etype];
+          const dgl_type_t dst_id = out_eid[etype];
+          CSRMatrix csr = vec_csr[etype];
+          NDArray ufeat = (vec_ufeat.size() == 0) ? NullArray() : vec_ufeat[src_id];
+          NDArray efeat = (vec_efeat.size() == 0) ? NullArray() : vec_efeat[etype];
+          NDArray out = vec_out[dst_id];
+          cpu::SpMMSumCsr<IdType, DType, Op>(bcast, csr, ufeat, efeat, out);
+        }
+      });
+    });
+  } 
+  else if (reduce == "max" || reduce == "min") {
+    SWITCH_BITS(bits, DType, {
+      SWITCH_OP(op, Op, {
+        /* Call  SpMM for each relation type */
+        for (dgl_type_t etype = 0; etype < ufeat_eid.size(); ++etype) {
+          const dgl_type_t src_id = ufeat_eid[etype];
+          const dgl_type_t dst_id = out_eid[etype];
+          CSRMatrix csr = vec_csr[etype];
+          NDArray ufeat = (vec_ufeat.size() == 0) ? NullArray() : vec_ufeat[src_id];
+          NDArray efeat = (vec_efeat.size() == 0) ? NullArray() : vec_efeat[etype];
+          NDArray out = vec_out[dst_id];
+          if (reduce == "max")
+            cpu::SpMMCmpCsr<IdType, DType, Op, cpu::op::Max<DType>>(
+                bcast, csr, ufeat, efeat, out, out_aux[0], out_aux[1]);
+          else
+            cpu::SpMMCmpCsr<IdType, DType, Op, cpu::op::Min<DType>>(
+                bcast, csr, ufeat, efeat, out, out_aux[0], out_aux[1]);
```

Comment:
it looks the code in the two branches are most or less the same. why do we not use if condition to check `reduce` to decide the template arguments?

Reply:
We can. I followed the original code structure.